### ISSUE TITLE
Pipeline fails to deploy changes to an application

### ISSUE THREAD

**Issue by sthaha** [openshiftio/openshift.io#2505]
### Steps
1. create a quickstart vertx (tested only using vertx template) project
1. choose `build` `->` `stage` `->` `approve` `->` `run`
1. wait for the pipeline to finish
1. Create a PR to the github project (make some changes to html for easy verification)  and merge it
1. Wait for pipeline to start and deploy the new version
1. Check if the new changes made in the PR is deployed

### Expected Result
1. Must see the new version deployed

### Actual Result
1. You would find that the version deployed (to `stage` and `run`) is still the initial version

### Preliminary investigation
@hrishin and I have been investigating this issue and we find that the `f-m-p` plugins when building somehow picks up the old version of the jar-file instead of the newer version.

### Progress

- [x] Fix bug in FMP   see: [PR #1225][fix-fmp]
- [x] Release FMP
- [x] Update qucikstart templates
- [x] Notify users to update FMP vesrion or patch the FMP version from pipeline flow

#### Vertx Boosters
- [x]  HTTP Booster - https://github.com/openshiftio-vertx-boosters/vertx-http-booster
- [x] Health check Booster - https://github.com/openshiftio-vertx-boosters/vertx-health-checks-booster

#### Springboot
- [x] HTTP Booster - https://github.com/snowdrop/spring-boot-http-booster
- [x] Health check Booster - https://github.com/snowdrop/spring-boot-health-check-booster

#### Wildfly swarm boosters
- [x] Http Booster - https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-rest-http
- [x] Health check Booster - https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-health-check

[fix-fmp]: https://github.com/fabric8io/fabric8-maven-plugin/pull/1225
[so-version]: https://stackoverflow.com/a/1172371

**Comment by hrishin** [comment]
The issue is correct image tag is not getting reffered:

As part of build step, Image is getting pushed into Jenkins namespace (username-jenkins) and deploymentconfig is referring the image SHA from stage, run (username-stage, username-run) namespaces. FMP queries the ImagesStreams and try to find the latest version of Image tag that is being pushed. But its logic of picking the correct ImageStream tag is incorrect. Hence it's deploying the only the first image.

**Comment by sthaha** [comment]
### Status Update

@hrishin has a fix for the [`fabric8-maven-plugin (f-m-p)`][fix-fmp] and would make new release of the f-m-p which includes the fix soon but it wouldn't solve the problem for everyone as the `pom.xml` in users repo would still point to the old version of f-m-p. For a complete solution

1. Users would have to update their pom.xml  to use `3.5.38` (going to be released) version of f-m-p

```xml
          <plugin>
            <groupId>io.fabric8</groupId>
            <artifactId>fabric8-maven-plugin</artifactId>
            <version>3.5.38</version>   <!-- NOTE: 3.5.28 at present -->
```

2. We would have to update the quickstart repos to this version. It is also worth considering using an [open-ended version][so-version] range like  `<version>[3.5.38,4.0.0) </version>`  so that a newer version of the plugin will be used when available. 

[fix-fmp]: https://github.com/fabric8io/fabric8-maven-plugin/pull/1225
[so-version]: https://stackoverflow.com/a/1172371

**Comment by chmouel** [comment]
@krishnapaparaju Srikrishna would your team be able to help us for updating those quickstarts?

**Comment by rohanKanojia** [comment]
**Wildfly swarm boosters**
Healthcheck Booster - https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-health-check
Crud Booster - https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-rest-http-crud
Http Booster - https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-rest-http
Configmap Booster - https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-configmap-redhat

**Comment by pbergene** [comment]
@krishnapaparaju any update on this Sev-1?

**Comment by chmouel** [comment]
Is it actually closed? we still need to fix the boosters, right?

**Comment by piyush1594** [comment]
I merged the PR there, it automatically got closed here, I also reopened it. Yaa we need to fix boosters.

**Comment by piyush1594** [comment]
Update : [FMP 3.5.38](https://repo.maven.apache.org/maven2/io/fabric8/fabric8-maven-plugin/3.5.38/) has been released.

**Comment by hrishin** [comment]
@sthaha @ppalaga @rohanKanojia 
One strange thing is https://github.com/openshiftio-vertx-boosters/vertx-http-booster/blob/master/pom.xml#L140 use FMP v.3.5.33
Then why OSIO VertX booster is having a different version of FMP ?

**Comment by maxandersen** [comment]
Are we 100% sure this is the actual issue ? wouldn't this imply that updates never worked before ? 
or is this because s2i wasn't used until recently ?

**Comment by hrishin** [comment]
> wouldn't this imply that updates never worked before?

It might have worked before if GET `/imagestreams` API returned the latest build tag as the first item in `imagestreams` list. Somehow if this behavior got changed then FMP is picking the wrong build. The code has been written based on this assumption in FMP.

>  Is this because s2i wasn't used until recently?

Not sure about it unless configuration changed explicitly. FMP does S2I build by default for OpenShift environment.

**Comment by krishnapaparaju** [comment]
@pbergene @chmouel missed out this GitHub issue, look into this and update this issue..

**Comment by hrishin** [comment]
### Status Update

1. _Users would have to update their pom.xml  to use `3.5.38` (going to be released) version of f-m-p:_
Work for updating boosters templates is in progress. Submitted the [PR](https://github.com/openshiftio/booster-parent/pull/26) to update f-m-p version in one place.  we need to update [this](https://github.com/openshiftio/booster-parent/pull/26) change references across all boosters application. @animuk is taking it further.
This approach not only requires to update all boosters applications templates but also existing user's applications.     

2. _Update the pipeline workflow to use right version of f-m-p:_
Instead of updating booster templates or existing applications, let's pass f-m-p version to `mvn` command like `mvn -Dfabric8-maven-plugin.version=3.5.28`  through pipeline build only.
This way neither new booster templates not existing applications `pom.xml` would require changes.  
We have tested this solution for vertx and springboot applications and its working. 
PR: https://github.com/fabric8io/fabric8-pipeline-library/issues/391
Jenkins logs for VertX: https://gist.github.com/hrishin/9ba9ea33096ee774e0521163ceaba968#file-vertx-jenkins-logs-L105
Jenkins logs for Springboot: https://gist.github.com/hrishin/ee5bbe868d24a199e43748e0c0d00af0#file-springbott-jenkins-L104

cc: @sthaha @maxandersen  @aslakknutsen @krishnapaparaju

**Comment by animuk** [comment]
Vertx Boosters
 HTTP Booster - PR Merged
 Health check Booster - already updated to version 20 for launcher-parent

Springboot

 HTTP Booster - already updated to version 20 for launcher-parent
 Health check Booster - already updated to version 20 for launcher-parent

Wildfly swarm boosters

 Http Booster - already a part of another PR for updating the version 20 for launcher-parent
 Health check Booster - already updated to version 20 for launcher-parent

**Comment by sthaha** [comment]
@hrishin 
> PR: fabric8io/fabric8-pipeline-library#391
seems to point to an issue rather than a PR

**Comment by hrishin** [comment]
>> PR: fabric8io/fabric8-pipeline-library#391
seems to point to an issue rather than a PR

@sthaha 
updated the comment :+1: 
PR: https://github.com/fabric8io/fabric8-pipeline-library/pull/392

**Comment by chmouel** [comment]
@sthaha  @animuk can we link all the GH issues that fix the boosters here ?

**Comment by animuk** [comment]
As per Hrishikesh, Vertx and Spring Boot is working fine with updated version. Only wildfly-swarm dont have updated fmp version ie. booster-parent version is still 10, instead of latest version 20. we are checking this.

**Comment by animuk** [comment]
As per @Ladicek 's comments, Wildfly swarm boosters are not updated till now. they are working for it.

**Comment by hrishin** [comment]
> As per Hrishikesh, Vertx and Spring Boot is working fine with updated version. Only wildfly-swarm dont have updated fmp version ie. booster-parent version is still 10, instead of latest version 20. we are checking this.

Now wildfly swarm also has updated FMP v3.5.28. It's in projects pom.xml only.

### Status Update
All 6 booster templates have updated with FMP v.3.5.28.

**Comment by hrishin** [comment]
### Status Update

Following tasks are completed

- [x] Fix bug in FMP : 
- [x] Release FMP : v3.5.38
- [x] Update qucikstart templates : 6 boosters applications

_Now user can deploy the application changes for fresh deployments in openshift.io._

Remaining task is 
- [ ] Patch existing users application for upgrading FMP to v3.5.38  either by
- Sending an issue report to users by mattermost/email and ask them to update `pom.xml`.
- Patching the `pom.xml` on the go through pipeline workflow.  PR: https://github.com/fabric8io/fabric8-pipeline-library/pull/392

cc: @sthaha @pradeepto @chmouel @maxandersen @kbsingh @aslakknutsen @stevengutz

**Comment by sthaha** [comment]
@hrishin 
> Sending an issue report to users by mattermost/email and ask them to update pom.xml.

Is that needed? wouldn't patching the pom.xml fix the issue?

**Comment by pbergene** [comment]
@sthaha that would fix the issue for any users that are new on the service or do a fresh quickstart, but not for those who have the pom.xml in their repositories already.

**Comment by sthaha** [comment]
@pbergene by patching the pom.xml I meant what the commit does here - https://github.com/hrishin/fabric8-pipeline-library/commit/07a8e134e7a0f225054a2ad2235a3bba97d69ffc which actually replaces the value in pom.xml with the latest version of f-m-p before running `mvn` which IIUC is sufficient to solve the issue.

**Comment by pbergene** [comment]
@sthaha OIC, sed to the rescue. I am reducing this to a Sev-2

**Comment by sthaha** [comment]
@pbergene please note that https://github.com/fabric8io/fabric8-pipeline-library/pull/392 isn't merged yet so the issue isn't solved yet in prod.

**Comment by hrishin** [comment]
> Is that needed? wouldn't patching the pom.xml fix the issue?

https://github.com/hrishin/fabric8-pipeline-library/commit/07a8e134e7a0f225054a2ad2235a3bba97d69ffc Its will fix the issue. The side effect of hard-coding FMP version will always override the applications FMP version and the user may feel annoying "why pipeline is using a different version than from `pom.xml`".

It's better to enhance the patching strategy like
```
if(application FMP version is older than 3.5.38) then
  patchFMPVersion
```

@sthaha

**Comment by pbergene** [comment]
Oh.. what's the ETA on that?

**Comment by hrishin** [comment]
### Status Update:

Working on a Remaining task
- [ ] Patch existing users application for upgrading FMP to v3.5.38

- WIP for PR: fabric8io/fabric8-pipeline-library#392, have done most of the changes required for patching 'pom.xml' but need sometime for testing. ETA for this  Monday. 
- Preparing an incident report on this issue.

**Comment by hrishin** [comment]
### Status Update:

Working on a remaining task
- [ ] Patch existing users application for upgrading FMP to v3.5.38

- WIP for PR: fabric8io/fabric8-pipeline-library#392. Making this patch work is becoming bit tricky. Any changes made for patching the `pom.xml` is causing unwanted changes to '`pom.xml` e.g. dropping <packaging> tag. Need bit more work on patching.

**Comment by hrishin** [comment]
blocked on #2721 for testing the patching mechanism

**Comment by hrishin** [comment]
### Status Update

Working on a remaining task
- [ ] Patch existing users application for upgrading FMP to v3.5.38 PR : fabric8io/fabric8-pipeline-library#392 . Tested this pipeline change and it works well.
- [ ] [WIP] Prepare an incident report

**Comment by hrishin** [comment]
### Status Update

Working on a remaining task
- [ ] Patch existing users application for upgrading FMP to v3.5.38 PR : fabric8io/fabric8-pipeline-library#392 . Tested this pipeline change and it works well.
- [ ] Prepare an incident report
